### ISSUE TITLE
slate-cli dependencies are not locked and could break unexpectedly

### ISSUE THREAD

**Issue by t-kelly** [Shopify/slate#238]
***tl;dr:** Applications, such as command line clients (aka `slate-cli`), can benefit from using npm shrinkwrap to secure dependencies and avoid spontaneous breakage when updating to the latest version.*

## Problem
1. Our current approach of using specific version numbers (e.g. `"gulp": "3.9.1"`) misses out on the benefits of semantic versioning. We lock people who our libraries to a specific version of a dependency, and don't allow them to get potentially beneficial, non-breaking changes (e.g. `"gulp": "3.9.2" or "gulp": "3.10.0"). 

2. In addition, specifying a specific dependency version works for our package's dependencies, but does not do anything to prevent their dependencies from breaking, e.g. `"gulp": "3.9.1"` has a dependency in it's `package.json` of `"eslint": "^1.7.3"`; eslint will still change as new versions are released.

### Where is this coming from?
In response to https://github.com/Shopify/slate/issues/160, I started checking which dependencies in our packages were up to date and how to best update them. From this, I've spent some time wrapping my head around specifying versions in `package.json`, and how  [`lock`](https://docs.npmjs.com/files/package-lock.json) files benefit projects. What follows is a summary of this research and what I think could change with Slate.

### package-lock.json
`package-lock.json` is a new feature in NPM 5 which is automatically generated by NPM and should be committed to version control. It is an exhaustive list that specifies the exact version of your project's dependencies, and their dependencies, and their dependencies, etc. 

`package-lock.json` is never published to NPM and is never used by other users who install your package for their project, e.g `npm install @shopify/slate-tools`. The only benefit `package-lock.json` provides is to developers who clone your projects repo and run `npm install`. It is a way of ensuring that you always have the same dev setup as every other dev. The yarn team [wrote a nice article](https://yarnpkg.com/blog/2016/11/24/lockfiles-for-all/) that explains this more.

### npm shrinkwrap
`npm-shrinkwrap.json` serves the same purpose as `package-lock.json`, except it **is published** to npm and it **is used when a user installs your package from NPM**. `npm-shrinkwrap.json` guarantees that the packages dependencies and their dependencies remain identical to the time of publishing.

`npm-shrinkwrap.json` is beneficial for applications, i.e. code that is not a dependency of other projects, because you want that application to always work when it's installed. `slate-cli` is an example of an application

`npm-shrinkwrap.json` is **NOT** beneficial for libraries, i.e. code that is a dependency of other projects. `slate-theme` and `slate-tools` are an example of libraries. Users of these libraries should be able to take advantage of the benefits of semantic versioning and download the latest, non-breaking version. 

## Proposed Solution
1. **Move `commands` found in `slate-tools` into `slate-cli`.** [Nicolás Bevacqua](https://ponyfoo.com/articles/immutable-npm-dependencies) makes a good point that CLI and its API should remain separate. `slate-tools` is currently a mix of API and CLI. Let's clean this up.
2. **Generate `npm-shrinkwrap.json` for `slate-cli` on [`prepublishOnly`](https://docs.npmjs.com/misc/scripts) and then delete it on [`postpublish`](https://docs.npmjs.com/misc/scripts).** That way we don't need to deal with npm-shrinkwrap when developing, it's only created and used for when we publish `slate-cli` to NPM. Also, `npm-shrinkwrap.json` should then always mimic what is set in our `package-lock.json`.
3. **Change `package.json` dependencies to use `^` instead of specific versions.**

**Comment by t-kelly** [comment]
[Slate v1.x ](https://github.com/Shopify/slate/tree/1.x)moves away from a global CLI. This is not a problem for future Slate versions.
